base_projects/Drivetrain/src/main/cpp/Robot.cpp

```diff
@@ -0,0 +1,367 @@
+        m_json.at("secondary encoder inverted").get<bool>();
+    m_cpr = m_json.at("cpr").get<double>();
+
+    std::string gyroType = m_json.at("gyro").get<std::string>();
+    wpi::StringRef gyroCtor = m_json.at("gyro ctor").get<std::string>();
+
+    std::cout << "Setup motors" << std::endl;
+    for (size_t i = 0; i < leftPorts.size(); i++) {
+      AddMotorController(leftPorts[i], controllerNames[i],
+                         leftMotorsInverted[i], &m_leftControllers);
+      AddMotorController(rightPorts[i], controllerNames[i],
+                         rightMotorsInverted[i], &m_rightControllers);
+    }
+
+    std::cout << "Setup encoders" << std::endl;
+    SetupEncoders(controllerNames[0], m_leftControllers.at(0).get(),
+                  leftEncoderInverted, leftEncoderPorts, m_leftCancoder,
+                  m_leftCANEncoder, m_leftEncoder, m_leftPosition, m_leftRate);
+    SetupEncoders(controllerNames[0], m_rightControllers.at(0).get(),
+                  rightEncoderInverted, rightEncoderPorts, m_rightCancoder,
+                  m_rightCANEncoder, m_rightEncoder, m_rightPosition,
+                  m_rightRate);
+
+    std::cout << "setup gyro" << std::endl;
+    if (gyroType == "Pigeon") {
+      std::string port_str = "";
+      if (gyroCtor.contains("WPI_TalonSRX")) {
+        port_str = std::string{gyroCtor[gyroCtor.find("-") + 1]};
+      } else {
+        port_str = gyroCtor;
+      }
+      // converts gyroCtor into port #
+      std::stringstream ss(port_str);
+      int srx_port = 0;
+      ss >> srx_port;
+      if (gyroCtor.contains("WPI_TalonSRX")) {
+        m_pigeon = std::make_unique<PigeonIMU>(new WPI_TalonSRX(srx_port));
+      } else {
+        m_pigeon = std::make_unique<PigeonIMU>(srx_port);
+      }
+
+      // setup functions
+      m_gyroPosition = [&, this] {
+        double xyz[3];
+        m_pigeon->GetAccumGyro(xyz);
+        return xyz[2];
+      };
+
+      m_gyroRate = [&, this] {
+        double xyz_dps[3];
+        m_pigeon->GetRawGyro(xyz_dps);
+        units::degrees_per_second_t rate{xyz_dps[2]};
+        return units::radians_per_second_t{rate}.to<double>();
+      };
+
+    } else {
+      if (gyroType == "ADXRS450") {
+        if (gyroCtor == "SPI.kMXP") {
+          m_gyro = std::make_unique<frc::ADXRS450_Gyro>(frc::SPI::Port::kMXP);
+        } else {
+          m_gyro =
+              std::make_unique<frc::ADXRS450_Gyro>(frc::SPI::Port::kOnboardCS0);
+        }
+        // } else if (gyroType == "NavX") {
+        //   if (gyroCtor == "SerialPort.kUSB") {
+        //     m_gyro = std::make_unique<AHRS>(frc::SerialPort::Port::kUSB);
+        //   } else if (gyroCtor == "I2C") {
+        //     m_gyro = std::make_unique<AHRS>(frc::I2C::Port::kMXP);
+        //   } else if (gyroCtor == "SerialPort.kMXP") {
+        //     m_gyro = std::make_unique<AHRS>(frc::SerialPort::Port::kMXP);
+        //   } else {
+        //     m_gyro = std::make_unique<AHRS>(frc::SPI::Port::kMXP);
+        //   }
```

Comment:
Should this be removed?

Reply:
Maybe for now? The issue is that NavX doesn't have Linux or mac binaries so the integration tests fail for it, but ideally we should be able to support it for the real version.

Reply:
I guess we could mark it as `FIXME: waiting on Linux and macOS builds for navX AHRS` to save us work later.